### PR TITLE
Unify response validation

### PR DESCRIPTION
**Pull request by jean-edouard-boulanger**
As promised, this is a followup [this PR](https://github.com/0b01001001/spectree/pull/338) which unifies and simplifies response validation across plugins. Response validation has been generalised in a single/simple function used across plugins.

### CODE REVIEW

File: spectree/plugins/base.py [hunk 1]
```diff
@@ -122,3 +125,65 @@ def get_func_operation_id(self, func: Callable, path: str, method: str):
...
+            # Make it possible to return an instance of the model __root__ type
+            # (i.e. not the root model itself).
+            try:
+                response_payload = validation_model(__root__=response_payload)
+            except ValidationError:
+                raise
+            else:
+                skip_validation = True
+                final_response_payload = serialize_model_instance(response_payload)
+        elif isinstance(response_payload, validation_model):
```
Comment (kemingy) [inline spectree/plugins/base.py:175]: Shall we use:

```py
if isinstance(response_payload, validation_model):
    ....
elif is_root_model(validation_model)
```

File: spectree/plugins/flask_plugin.py [hunk 2]
```diff
@@ -206,63 +201,38 @@ def validate(
 
         status = 200
         rest = []
-        if resp and isinstance(result, tuple) and isinstance(result[0], BaseModel):
```
[the remark this answers is not in the document]
Comment (kemingy) [reply spectree/plugins/flask_plugin.py:209]: Agree.

File: spectree/plugins/starlette_plugin.py [hunk 3]
```diff
@@ -141,13 +136,14 @@ async def validate(
...
-                    resp_validation_error = err
-                    response = JSONResponse(err.errors(), 500)
+            try:
+                validate_response(
+                    skip_validation=skip_validation,
+                    validation_model=resp.find_model(response.status_code),
+                    response_payload=RawResponsePayload(payload=response.body),
+                )
+            except ValidationError as err:
+                response = JSONResponse(err.errors(), 500)
```
[the remark this answers is not in the document]
Comment (kemingy) [reply spectree/plugins/starlette_plugin.py:146]: Nice catch. Yeah, we can do it in another PR.

File: spectree/plugins/base.py [hunk 4]
```diff
@@ -122,3 +125,65 @@ def get_func_operation_id(self, func: Callable, path: str, method: str):
...
+
+
+@dataclass(frozen=True)
+class RawResponsePayload:
+    payload: Union[JsonType, bytes]
+
+
+@dataclass(frozen=True)
+class ResponseValidationResult:
+    validated_response_payload: Any
```
Comment (kemingy) [inline spectree/plugins/base.py:137]: maybe `payload` is enough since the outer class name already contains "validation"

**Review by kemingy — COMMENTED** [review]
Impressive work! 
Thanks for your contribution! 🎉

**Comment by jean-edouard-boulanger** [comment]
Thank you for taking the time to review 🙌

**Review by kemingy — APPROVED** [review]

**Pull request merged**